src/include/common/worker_pool.h

```diff
@@ -0,0 +1,159 @@
+ * shutdown. Calls to Startup() and Shutdown() are thread-safe.
+ */
+class WorkerPool {
+ public:
+  /**
+   * Initialize the worker pool. Once the number of worker is set by the constructor
+   * it cannot be changed.
+   *
+   * After initialization of the worker pool with a given task queue, worker threads
+   * do NOT start running. Need call StartUp() to start working on tasks.
+   *
+   * @param pool_name the name of the worker pool
+   * @param num_workers the number of workers in this pool
+   * @param task_queue a queue of tasks
+   */
+  WorkerPool(std::string pool_name, uint32_t num_workers, TaskQueue task_queue)
+      : pool_name_(std::move(pool_name)),
+        num_workers_(num_workers),
+        is_running_(false),
+        task_queue_(std::move(task_queue)) {}
+
+  /**
+   * Destructor. Wake up all workers and let them finish before it's destroyed.
+   */
+  ~WorkerPool() {
+    std::unique_lock<std::mutex> lock(task_lock_);  // grab the lock
+    is_running_ = false;                            // signal all the threads to shutdown
+    task_cv_.notify_all();                          // wake up all the threads
+    lock.unlock();                                  // free the lock
+    for (auto &thread : workers_) thread.join();
+  }
+
+  /**
+   * Start the worker pool. If the task queue is empty or we run out of tasks,
+   * workers will be put into sleep.
+   */
+  void Startup() {
+    is_running_ = true;
+    while (workers_.size() < num_workers_) {
+      AddThread();
+    }
+  }
+
+  /**
+   * It tells all worker threads to finish their current task and stop working.
+   * No more tasks will be consumed. It waits until all worker threads stop working.
+   */
+  void Shutdown() {
+    is_running_ = false;
+    std::unique_lock<std::mutex> lock(task_lock_);
+    // wait for all the threads to finish
+    finished_cv_.wait(lock, [this] { return busy_workers_ == 0; });
+  }
+
+  /**
+   * Add a task to the task queue and inform worker threads.
+   *
+   * Side effect: If the pool has been shutdown, calling SubmitTask will automatically
+   * re-startup the pool and resume to work.
+   * @param func the new task
+   */
+  template <typename F>
+  inline void SubmitTask(const F &func) {
+    if (!is_running_) {
+      Startup();
+    }
+    std::unique_lock<std::mutex> lock(task_lock_);  // grab the lock
+    task_queue_.emplace(std::move(func));
+    task_cv_.notify_one();
+  }
+
+  /**
+   * Block until all the tasks in the task queue has been completed
+   */
+  void WaitUtilFinish() {
+    std::unique_lock<std::mutex> lock(task_lock_);
+    // wait for all the tasks to complete
+    finished_cv_.wait(lock, [this] { return busy_workers_ == 0 && task_queue_.empty(); });
+  }
+
+  /**
+   * Get the number of worker threads in this pool
+   *
+   * @return The number of worker threads
+   */
+  uint32_t NumWorkers() const { return num_workers_; }
+
+ private:
+  // The name of this pool
+  std::string pool_name_;
```

Comment:
This seems very unnecessary. Either way there's not even an accessor for it so it is unusable anyway. We can remove it for now.